src/coreclr/src/debug/dbgutil/elfreader.h

```diff
@@ -0,0 +1,86 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include <windows.h>
+#include <clrdata.h>
+#include <cor.h>
+#include <cordebug.h>
+#define __STDC_FORMAT_MACROS
+#include <inttypes.h>
+#include <elf.h>
+#ifdef HOST_UNIX
+#include <link.h>
+#endif // HOST_UNIX
+#include <string>
+#include <vector>
+
+#if TARGET_64BIT
+#define PRIx PRIx64
+#define PRIu PRIu64
+#define PRId PRId64
+#define PRIA "016"
+#define PRIxA PRIA PRIx
+#define TARGET_WORDSIZE 64
+#else
+#define PRIx PRIx32
+#define PRIu PRIu32
+#define PRId PRId32
+#define PRIA "08"
+#define PRIxA PRIA PRIx
+#define TARGET_WORDSIZE 32
+#endif
+
+#ifndef ElfW
+/* We use this macro to refer to ELF types independent of the native wordsize.
+   `ElfW(TYPE)' is used in place of `Elf32_TYPE' or `Elf64_TYPE'.  */
+#define ElfW(type)      _ElfW (Elf, TARGET_WORDSIZE, type)
+#define _ElfW(e,w,t)    _ElfW_1 (e, w, _##t)
+#define _ElfW_1(e,w,t)  e##w##t
+#endif
+
+typedef struct {
+    int32_t BucketCount;
+    int32_t SymbolOffset;
+    int32_t BloomSize;
+    int32_t BloomShift;
+} GnuHashTable;
+
+class ElfReader
+{
+private:
+    void* m_rdebugAddr;                     // DT_DEBUG
+    void* m_gnuHashTableAddr;               // DT_GNU_HASH
+    void* m_stringTableAddr;                // DT_STRTAB
+    int m_stringTableSize;                  // DT_STRSIZ
+    void* m_symbolTableAddr;                // DT_SYMTAB
+
+    GnuHashTable m_hashTable;               // gnu hash table info
+    int32_t* m_buckets;                     // gnu hash table buckets    
+    void* m_chainsAddress;
+
+public:
+    ElfReader();
+    virtual ~ElfReader();
+    bool PopulateELFInfo(uint64_t baseAddress);
+    bool PopulateELFInfo(ElfW(Phdr)* phdrAddr, int phnum);
+    bool TryLookupSymbol(std::string symbolName, uint64_t* symbolOffset);
+#ifdef HOST_UNIX
+     bool EnumerateLinkMapEntries();
+#endif
+    bool EnumerateProgramHeaders(uint64_t baseAddress, ElfW(Dyn)** pdynamicAddr);
+    bool EnumerateProgramHeaders(ElfW(Phdr)* phdrAddr, int phnum, uint64_t baseAddress, ElfW(Dyn)** pdynamicAddr);
+
+private:
+    bool GetSymbol(int32_t index, ElfW(Sym)* symbol);
+    bool InitializeGnuHashTable();
+    bool GetPossibleSymbolIndex(const std::string& symbolName, std::vector<int32_t>& symbolIndexes);
+    uint32_t Hash(const std::string& symbolName);
+    bool GetChain(int index, int32_t* chain);
+    bool GetStringAtIndex(int index, std::string& result);
+    bool EnumerateDynamicEntries(ElfW(Dyn)* dynamicAddr);
+    virtual void ForEachModule(uint64_t baseAddress, std::string& moduleName) { };
```

Comment:
This one looks like a visitor or an event so typical names might be VisitModule() or OnReadModule()